dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/AbstractTrackerValidationHook.java

```diff
@@ -0,0 +1,274 @@
+            }
+
+            FeatureType typeFromName = FeatureType.getTypeFromName( geometry.getGeometryType() );
+
+            if ( FeatureType.NONE == featureType || featureType != typeFromName )
+            {
+                errorReporter.addError( newReport( TrackerErrorCode.E1012 )
+                    .addArg( featureType.name() ) );
+            }
+        }
+    }
+
+    protected OrganisationUnit getOrganisationUnit( TrackerBundle bundle, TrackedEntity te )
+    {
+        Objects.requireNonNull( bundle, TRACKER_BUNDLE_CANT_BE_NULL );
+        Objects.requireNonNull( te, TRACKED_ENTITY_CANT_BE_NULL );
+
+        TrackedEntityInstance trackedEntityInstance = PreheatHelper
+            .getTrackedEntityInstance( bundle, te.getTrackedEntity() );
+
+        OrganisationUnit organisationUnit =
+            trackedEntityInstance != null ? trackedEntityInstance.getOrganisationUnit() : null;
+
+        return bundle.getImportStrategy().isCreate()
+            ? PreheatHelper.getOrganisationUnit( bundle, te.getOrgUnit() )
+            : organisationUnit;
+    }
+
+    protected TrackedEntityType getTrackedEntityType( TrackerBundle bundle, TrackedEntity te )
+    {
+        Objects.requireNonNull( bundle, TRACKER_BUNDLE_CANT_BE_NULL );
+        Objects.requireNonNull( te, TRACKED_ENTITY_CANT_BE_NULL );
+
+        TrackedEntityInstance trackedEntityInstance = PreheatHelper
+            .getTrackedEntityInstance( bundle, te.getTrackedEntity() );
+
+        TrackedEntityType trackedEntityType =
+            trackedEntityInstance != null ? trackedEntityInstance.getTrackedEntityType() : null;
+
+        return bundle.getImportStrategy().isCreate()
+            ? PreheatHelper.getTrackedEntityType( bundle, te.getTrackedEntityType() )
+            : trackedEntityType;
+    }
+
+    protected ProgramInstance getProgramInstance( User actingUser, ProgramInstance programInstance,
+        TrackedEntityInstance trackedEntityInstance, Program program )
+    {
+        Objects.requireNonNull( program, PROGRAM_INSTANCE_CANT_BE_NULL );
+        Objects.requireNonNull( actingUser, USER_CANT_BE_NULL );
+
+        if ( program.isRegistration() )
+        {
+            if ( programInstance == null && trackedEntityInstance != null )
+            {
+                List<ProgramInstance> activeProgramInstances = new ArrayList<>( programInstanceService
+                    .getProgramInstances( trackedEntityInstance, program, ProgramStatus.ACTIVE ) );
+
+                if ( activeProgramInstances.size() == 1 )
+                {
+                    programInstance = activeProgramInstances.get( 0 );
+                }
+            }
+        }
+        else
+        {
+            // NOTE: This is cached in the prev. event importer? What do we do here?
+            List<ProgramInstance> activeProgramInstances = programInstanceService
+                .getProgramInstances( program, ProgramStatus.ACTIVE );
+
+            if ( activeProgramInstances.isEmpty() )
```

Comment:
Shouldn't be needed anymore.. we do this in the metadata post create/update hooks